merkle-tree/hasher/src/bytes.rs

```diff
@@ -0,0 +1,217 @@
+/// are used inside compressed accounts.
+pub trait AsByteVec {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>>;
+}
+
+macro_rules! impl_as_byte_vec_for_integer_type {
+    ($int_ty:ty) => {
+        impl AsByteVec for $int_ty {
+            fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+                vec![self.to_ne_bytes().to_vec()]
+            }
+        }
+    };
+}
+
+macro_rules! impl_as_byte_vec_for_primitive_type {
+    ($int_ty:ty) => {
+        impl AsByteVec for $int_ty {
+            fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+                let len = mem::size_of_val(self);
+                let self_ptr: *const Self = self;
+                // SAFETY:
+                // - All the primitive types we implement this macro for have
+                //   an exact size (`len`). There is no chance of reads out of
+                //   bounds.
+                // - Casting Rust primitives to bytes works fine, there is no
+                //   chance of undefined behavior.
+                // - Unfortunately, there is no way to achieve the similar
+                //   result with fully safe code. If we tried to do anything
+                //   like `&self.to_ne_bytes()` or `self.to_ne_bytes().as_slice()`,
+                //   compiler would complain with "cannot return reference to
+                //   temporary value".
+                let self_byte_slice = unsafe { slice::from_raw_parts(self_ptr.cast::<u8>(), len) };
+                vec![self_byte_slice.to_vec()]
+            }
+        }
+    };
+}
+
+impl_as_byte_vec_for_integer_type!(i8);
+impl_as_byte_vec_for_integer_type!(u8);
+impl_as_byte_vec_for_integer_type!(i16);
+impl_as_byte_vec_for_integer_type!(u16);
+impl_as_byte_vec_for_integer_type!(i32);
+impl_as_byte_vec_for_integer_type!(u32);
+impl_as_byte_vec_for_integer_type!(i64);
+impl_as_byte_vec_for_integer_type!(u64);
+impl_as_byte_vec_for_integer_type!(isize);
+impl_as_byte_vec_for_integer_type!(usize);
+impl_as_byte_vec_for_integer_type!(i128);
+impl_as_byte_vec_for_integer_type!(u128);
+
+impl_as_byte_vec_for_primitive_type!(bool);
+
+impl<T> AsByteVec for Option<T>
+where
+    T: AsByteVec,
+{
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        match self {
+            Some(hashable) => {
+                let mut bytes = hashable.as_byte_vec();
+                bytes.reserve(1);
+                bytes.insert(0, vec![1]);
+                bytes
+            }
+            None => vec![vec![0]],
+        }
+    }
+}
+
+impl<const N: usize> AsByteVec for [u8; N] {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.to_vec()]
+    }
+}
+
+impl AsByteVec for String {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.as_bytes().to_vec()]
+    }
+}
+
+#[cfg(feature = "solana")]
+impl AsByteVec for solana_program::pubkey::Pubkey {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.to_bytes().to_vec()]
+    }
+}
+
+#[cfg(test)]
```

Comment:
would it make sense to add test cases for additional edge cases such as:
* string / empty string
* option<T> with none, some
* empty array

Reply:
Done